### PR TITLE
Upgrade to Quarkus 3.14.2

### LINKED ISSUE #32198: error message "Address already in use" should state which address/port in particular
### Description

On  keycloack 25.0.2
```
KEYCLOAK_ADMIN=admin KEYCLOAK_ADMIN_PASSWORD=admin bin/kc.sh start-dev
2024-08-16 13:42:03,529 INFO  [org.infinispan.CONTAINER] (ForkJoinPool.commonPool-worker-1) ISPN000556: Starting user marshaller 'org.infinispan.jboss.marshalling.core.JBossUserMarshaller'
2024-08-16 13:42:04,285 INFO  [org.keycloak.broker.provider.AbstractIdentityProviderMapper] (main) Registering class org.keycloak.broker.provider.mappersync.ConfigSyncEventListener
2024-08-16 13:42:04,366 INFO  [org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory] (main) Node name: node_625156, Site name: null
2024-08-16 13:42:05,298 INFO  [com.arjuna.ats.jbossatx] (main) ARJUNA032014: Stopping transaction recovery manager
2024-08-16 13:42:05,347 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (development) mode
2024-08-16 13:42:05,347 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Unable to start HTTP server
2024-08-16 13:42:05,348 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: java.lang.IllegalStateException: Unable to start the management interface
2024-08-16 13:42:05,348 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Unable to start the management interface
2024-08-16 13:42:05,348 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Address already in use
2024-08-16 13:42:05,348 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) For more details run the same command passing the '--verbose' option. Also you can use '--help' to see the details about the usage of the particular command.
```

the error `ERROR: Address already in use` I guess it's correct but it does not help me to find what in particular is failing. 

I eventually figured out that keycloak by default is using :8080 and :9000 so it had to be one of those.  So running `bin/kc.sh start-dev --http-management-port 9002` got rid of the problem. 

But I think the error message should be improve like  this 

```
2024-08-16 13:42:05,348 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Address already in use (0.0.0.0:9000) 

```

or 
```
2024-08-16 13:42:05,348  INFO [io.quarkus] (main) Trying management interface on http://0.0.0.0:9000.
2024-08-16 13:42:05,348 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Address already in use
```

### Discussion

_No response_

### Motivation

Well, it will save some time and confusion specially for beginners, that may not understand that keycloack is using other ports beside 8080.

### Details

_No response_

### LINKED ISSUE #32298: [quarkus-next] Keycloak CI failure: Cannot invoke "io.quarkus.runtime.Application.isStarted()" because "app" is null
### Description:
Integration tests fails due to the following:
```
jakarta.persistence.PersistenceException: [PersistenceUnit: keycloak-default] Unable to build Hibernate SessionFactory
at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.persistenceException(FastBootEntityManagerFactoryBuilder.java:129)
at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:89)
at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:72)
at jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:80)
at jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:55)
at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:154)
at io.quarkus.hibernate.orm.runtime.JPAConfig$1.run(JPAConfig.java:61)
at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: io.quarkus.runtime.configuration.ConfigurationException: Persistence unit 'keycloak-default' was configured to run with a database version of at least '2.3.232', but the actual version is '2.2.0'. Consider upgrading your database. Alternatively, rebuild your application with 'quarkus.datasource.db-version=2.2.0' (but this may disable some features and/or impact performance negatively).
at io.quarkus.hibernate.orm.runtime.service.QuarkusRuntimeInitDialectFactory.checkActualDbVersion(QuarkusRuntimeInitDialectFactory.java:70)
at io.quarkus.hibernate.orm.runtime.observers.QuarkusSessionFactoryObserverForDbVersionCheck.sessionFactoryCreated(QuarkusSessionFactoryObserverForDbVersionCheck.java:15)
at org.hibernate.internal.SessionFactoryObserverChain.sessionFactoryCreated(SessionFactoryObserverChain.java:35)
at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:323)
at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:87)
... 6 more
Exception in thread "Shutdown thread" java.lang.NullPointerException: Cannot invoke "io.quarkus.runtime.Application.isStarted()" because "app" is null
at io.quarkus.runtime.ApplicationLifecycleManager$ShutdownHookThread.run(ApplicationLifecycleManager.java:455)
06:59:53,649 ERROR [org.keycloak.testsuite.actions.AppInitiatedActionDeleteCredentialTest] [AppInitiatedActionDeleteCredentialTest] null()
```

The failure seems to be specific to `-Pauth-server-quarkus`.

Full log can be found here: 
https://github.com/keycloak/keycloak/actions/runs/10481101059

### Steps to reproduce:
```
mvn test -Pauth-server-quarkus -Dtest="org.keycloak.testsuite.actions.AppInitiatedActionDeleteCredentialTest#missingActionId" -pl testsuite/integration-arquillian/tests/base
```

### Additional information: 
The issue seems to be caused by https://github.com/quarkusio/quarkus/pull/42502, as the message from the log indicates:
```
Persistence unit 'keycloak-default' was configured to run with a database version of at least '2.3.232', but the actual version is '2.2.0'.
```
Therefore, H2 upgrade seems to be required.

### LINKED ISSUE #32517: Upgrade to Quarkus 3.14.2
### Description

https://github.com/quarkusio/quarkus/releases/tag/3.14.2

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by vmuzikar**
Closes #32517
Closes #32198 (edited by @mabartos)

Quarkus 3.14 introduced a bunch of (hidden) breaking changes.

### Oracle Driver extension
The extension now [touches](https://github.com/quarkusio/quarkus/commit/19d9ed1b269cafa8045dcf2b8b3db2dc4c290b89#diff-1744c4a730f826e51f66dd21c87ab13fc749bddbc011373540bcb2afd1cdd2ddL14) some of the driver classes in a recorder. This revealed how fragile is [our approach](https://github.com/keycloak/keycloak/blob/main/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/IgnoredArtifacts.java) with `quarkus.class-loading.removed-artifacts`. It turned out that option does not seem to be reflected until the server is reaugmented. What happens when the server is started without augmenting it first (i.e. when doing `kc.sh start --optimized`) is that the recorder is executed at the server start but it tries to use the DB driver classes that are not part of our dist.

The solution to that was to add a validation that needs to be built first. I believe this is reasonable as it's what we [document](https://github.com/keycloak/keycloak/blob/4b95b42590beb695ef0d794b8db6f0517fff7591/docs/guides/server/configuration.adoc?plain=1#L212) users need to do anyway.

Still, I find the approach to exclude a Quarkus extension using `quarkus.class-loading.removed-artifacts` very fragile and I'll open a discussion with the Quarkus team about it.

### Hibernate upgrade
The Hibernate upgrade [made it to be more strict and less forbidding](https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.14#upgrade-to-hibernate-orm-66) and as a result it throws exception in some cases where we were not perhaps following best practices.

This required some changes especially in how we handle composite roles.

### Liquibase upgrade
The Liquibase upgrade seems to have caused some issues around transactions. Added a workaround for that.

### Regression in H2 (unresolved yet)
See https://github.com/keycloak/keycloak/issues/32785

### CODE REVIEW

**Comment by mabartos** [comment]
@vmuzikar Please accommodate changes to the SmallRye Health config root (https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.14#smallrye-health).

Changing from `quarkus.health.extensions.enabled` => `quarkus.smallrye-health.extensions.enabled`

Thanks!

**Comment by mabartos** [comment]
@vmuzikar jfyi - I'm a little bit worried about the 409 Conflict flakiness. Could you please check them as well once you finish the work on this?

**Comment by vmuzikar** [comment]
> @vmuzikar jfyi - I'm a little bit worried about the 409 Conflict flakiness. Could you please check them as well once you finish the work on this?

This turned out to be a real issue hidden by our relaxed flaky test logic. Realms are sometimes not deleted properly.

**Comment by vmuzikar** [comment]
I was not able to track down the flakiness of HostnameV2Test yet. It's not really specific to that test, running it separately doesn't reproduce it, only as part of the testsuite. I also noticed it happened [before](https://github.com/keycloak/keycloak/pull/30590#issuecomment-2179494665) on other tests too, the Quarkus upgrade seems to just expose it more.

Since it's really just flakiness (e.g. in the last run it passed without a single failure), I believe it should be safe to proceed. I'd like to resolve it as a follow up to not hold the upgrade any longer. WDYT @mabartos?

**Comment by mabartos** [comment]
> I'd like to resolve it as a follow up to not hold the upgrade any longer. WDYT @mabartos?

Yes, definitely. Thanks

**Comment by mabartos** [comment]
@vmuzikar I have not looked at it much yet, but as the PR touches a lot of different stuff, it'd be good to track it somehow - creating separate issues and referencing them? Perhaps provide more information to the commit message as well.

**Comment by vmuzikar** [comment]
@mabartos I'm not sure if we should create more issues as they are not really relevant without the Quarkus upgrade. I tried to provide most details in the PR description. Does that work for you?

Once we're ready for the final review, I plan to tag other codeowners of the code I touched.

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java [hunk 1]
```diff
@@ -447,8 +447,19 @@ public boolean removeRole(RoleModel role) {
...
+        // At the same time, can't use the persist cascade type on the compositeRoles field because in that case
+        // we could not still use a native query as a different problem would arise - it may happen that a parent role that
+        // has this role as a composite is present in the persistence context. In that case it, the role would be re-created
+        // again after deletion through persist cascade type.
+        // So in any case, native query is not an option. This is not optimal as it executes additional queries but
+        // the alternative of clearing the persistence context is not either as we don't know if something currently present
+        // in the context is not needed later.
+        Stream<RoleEntity> parentRoles = em.createNamedQuery("getParentRolesOfACompositeRole", RoleEntity.class).setParameter("compositeRole", roleEntity).getResultStream();
+        parentRoles.forEach(parentRole -> parentRole.getCompositeRoles().remove(roleEntity));
+        parentRoles.close();
```
Comment (mabartos) [inline model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:461]: Isn't there any better approach how to achieve it? Perhaps ask someone from the Hibernate team?

**Review by mabartos — APPROVED** [review]
@vmuzikar Nice job! Quarkus side looks fine to me. As you've mentioned, it's necessary to involve more eyes on this - even maybe outside of our team.

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java (hunk 1 above)
Comment (vmuzikar) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:461]: I did not talked with the Hibernate team yet. Let's wait first for the core team's review. :)

**Comment by vmuzikar** [comment]
Spent some more time debugging the flaky hostname test and turned out to be much deeper issue: https://github.com/keycloak/keycloak/issues/32785.

Nevertheless, since it affects just H2, I'd like to still proceed with the PR. WDYT?

**Comment by vmuzikar** [comment]
Given the nature of the changes, adding more reviewers.

**Review by Pepo48 — APPROVED** [review]
I agree with merging this and sorting out the H2 issue as a follow-up. Thanks @vmuzikar for the relentless work that you've put into this. From my perspective it looks fine at it's current state.

**Review by ahus1 — APPROVED** [review]
LGTM. Thank you for those investigations. 

I read up on why we remove the user attributes and other the way we do it. I agree it works, still it looks quite non-standard. Still I don't have a better idea. 

I agree that `removed-artifacts` is probably an edge case for Quarkus, and it would simplify our code base not to use it. 

About the H2 update: I looked at the H2 GitHub repository and saw that there is no open issue that I would connect to this, and no changes after the release look to be connect to that. And the release we're using seems to be quite fresh. I suggest to open an issue with the H2 team with details you can share. Maybe they can recommend some additional logging on the H2 level.

**Comment by vmuzikar** [comment]
> I suggest to open an issue with the H2 team with details you can share. Maybe they can recommend some additional logging on the H2 level.

@ahus1 In the meantime, it turned out the issue is not reproducible with Undertow. That suggest rather Quarkus issue than H2.

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java (hunk 1 above)
Comment (mposolda) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java:461]: Don't have better idea either. Thanks for adding the comment as that can be useful for the future references.

**Review by mposolda — APPROVED** [review]
Approving. +1 to handle H2 issue as a follow-up.

Review from some DB guru (which I don't consider myself to be) won't hurt, but at the same time, it can be probably good to have this PR ASAP, so we have time to catch potential issues before the Keycloak 26.

**Review by shawkins — APPROVED** [review]
LGTM, thanks @vmuzikar

**Comment by vmuzikar** [comment]
Thank you for the reviews! I will go ahead and merge this later today as we should start testing with this soon to see if there are any hidden issues. We can do improvements as a follow-up.

**Review by vramik — APPROVED** [review]
LGTM, thank you @vmuzikar.

**Comment by vmuzikar** [comment]
Merged. Thank you all involved and please keep your fingers 🤞 it doesn't cause (too many) so far hidden regressions. :)

**Pull request merged**